# Building a Spam Filter with Naive Bayes

In this project, we're going to build a spam filter for SMS messages using the multinomial Naive Bayes algorithm. Our goal is to write a program that classifies new messages with an accuracy greater than 80% — so we expect that more than 80% of the new messages will be classified correctly as spam or ham (non-spam).

To train the algorithm, we'll use a dataset of 5,572 SMS messages that are already classified by humans. The dataset was put together by Tiago A. Almeida and José María Gómez Hidalgo, and it can be downloaded from [the The UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection). The data collection process is described in more details [on this page](http://www.dt.fee.unicamp.br/~tiago/smsspamcollection/#composition), where you can also find some of the papers authored by Tiago A. Almeida and José María Gómez Hidalgo.

## Exploring Data
Let's start by reading in the dataset.

In [1]:
# open data
import pandas as pd
sms_spam=pd.read_csv('SMSSpamCollection', sep='\t', header=None, names=['Label', 'SMS'])
sms_spam

Label                                                SMS
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [2]:
sms_spam['Label'].value_counts(normalize=True)

ham     0.865937
spam    0.134063
Name: Label, dtype: float64

We can see there are 5572 sms messages in data and  about 87% of the messages are ham, and the remaining 13% are spam.

## Training and Test Set
We're now going to split our dataset into a training and a test set, where the training set accounts for 80% of the data, and the test set for the remaining 20%.

In [3]:
# Randomize the dataset
data_randomize=sms_spam.sample(frac=1, random_state=1)

#Training and Test Split
training_test_index=round(len(data_randomize)*0.8)
training_set=data_randomize.iloc[:training_test_index].reset_index(drop=True)
test_set=data_randomize.iloc[training_test_index:].reset_index(drop=True)



In [4]:
# display training set
training_set

Label                                                SMS
0      ham                       Yep, by the pretty sculpture
1      ham      Yes, princess. Are you going to make me moan?
2      ham                         Welp apparently he retired
3      ham                                            Havent.
4      ham  I forgot 2 ask ü all smth.. There's a card on ...
...    ...                                                ...
4453   ham  Sorry, I'll call later in meeting any thing re...
4454   ham  Babe! I fucking love you too !! You know? Fuck...
4455  spam  U've been selected to stay in 1 of 250 top Bri...
4456   ham  Hello my boytoy ... Geeee I miss you already a...
4457   ham                           Wherre's my boytoy ? :-(

[4458 rows x 2 columns]

In [5]:
training_set['Label'].value_counts(normalize=True)

ham     0.86541
spam    0.13459
Name: Label, dtype: float64

In [6]:
#display test set
test_set.head(10)

Label                                                SMS
0   ham          Later i guess. I needa do mcat study too.
1   ham             But i haf enuff space got like 4 mb...
2  spam  Had your mobile 10 mths? Update to latest Oran...
3   ham  All sounds good. Fingers . Makes it difficult ...
4   ham  All done, all handed in. Don't know if mega sh...
5   ham  But my family not responding for anything. Now...
6   ham                                           U too...
7   ham  Boo what time u get out? U were supposed to ta...
8   ham  Genius what's up. How your brother. Pls send h...
9   ham                             I liked the new mobile

In [7]:
test_set['Label'].value_counts(normalize=True)

ham     0.868043
spam    0.131957
Name: Label, dtype: float64

We'll now analyze the percentage of spam and ham messages in the training and test sets. We expect the percentages to be close to what we have in the full dataset, where about 87% of the messages are ham, and the remaining 13% are spam.

## Basic theory of spam filter
On the previous screen, we split our dataset into a training set and a test set. The next big step is to use the training set to teach the algorithm to classify new messages.

our Naive Bayes algorithm will make the classification based on the results it gets to these two equations (note that we replaced "SpamC" with "Ham" inside the second equation below):
<img src="equation1.png" width="500" height="300">

Also, to calculate `P(wi|Spam)` and `P(wi|Ham)` inside the formulas above, recall that we need to use these equations:
<img src="equation2.png" width="500" height="300">

Let's also summarize what the terms in the equations above mean:
<img src="equation3.png" width="500" height="300">

## Clean Data
To calculate all these probabilities, we'll first need to perform a bit of data cleaning to bring the data in a format that will allow us to extract easily all the information we need. Right now, our training and test sets have this format (the messages are fictitious to make the example easier to understand):
<img src="table1.png" width="500" height="300">

To make the calculations easier, we want bring the data to this format (the table below is a transformation of the table you see above):
<img src="table2.png" width="500" height="300">

We'll begin with removing all the punctuation and bringing every letter to lower case.

In [8]:
# data review
training_set.head()

Label                                                SMS
0   ham                       Yep, by the pretty sculpture
1   ham      Yes, princess. Are you going to make me moan?
2   ham                         Welp apparently he retired
3   ham                                            Havent.
4   ham  I forgot 2 ask ü all smth.. There's a card on ...

In [9]:
# cleaning data
training_set['SMS']=training_set['SMS'].str.replace('\W',' ')
training_set['SMS']=training_set['SMS'].str.lower()
training_set.head()

Label                                                SMS
0   ham                       yep  by the pretty sculpture
1   ham      yes  princess  are you going to make me moan 
2   ham                         welp apparently he retired
3   ham                                            havent 
4   ham  i forgot 2 ask ü all smth   there s a card on ...

With the exception of the "Label" column, every other column in the transformed table above represents a unique word in our vocabulary (more specifically, each column shows the frequency of that unique word for any given message). Recall from the previous mission that we call the set of unique words a vocabulary.



### Vocabulary list
We'll eventually bring the training set to that format ourselves, but first, let's create a list with all of the unique words that occur in the messages of our training set.


In [10]:
training_set['SMS'] = training_set['SMS'].str.split()

vocabulary = []
for sms in training_set['SMS']:
    for word in sms:
        vocabulary.append(word)
        
vocabulary = list(set(vocabulary))
len(vocabulary)

7783

In the train set, there are 7783 unique words for our vocabulary.

### The Final Training Set
We're now going to use the vocabulary we just created to make the data transformation we want.

In [11]:
word_counts_per_sms = {unique_word: [0] * len(training_set['SMS']) for unique_word in vocabulary}

for index, sms in enumerate(training_set['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1

In [12]:
word_counts = pd.DataFrame(word_counts_per_sms)
word_counts.head()

chinky  remembr  goigng  humans  pc  erupt  laugh  history  ta  wannatell  \
0       0        0       0       0   0      0      0        0   0          0   
1       0        0       0       0   0      0      0        0   0          0   
2       0        0       0       0   0      0      0        0   0          0   
3       0        0       0       0   0      0      0        0   0          0   
4       0        0       0       0   0      0      0        0   0          0   

   ...  hont  somewheresomeone  220  nor  boytoy  09066612661  09061743386  \
0  ...     0                 0    0    0       0            0            0   
1  ...     0                 0    0    0       0            0            0   
2  ...     0                 0    0    0       0            0            0   
3  ...     0                 0    0    0       0            0            0   
4  ...     0                 0    0    0       0            0            0   

   paining  suffers  sleepin  
0        0        0        0  
1        0        0        0  
2        0        0        0  
3        0        0        0  
4        0        0        0  

[5 rows x 7783 columns]

In [13]:
training_set_clean=pd.concat([training_set,word_counts],axis=1)
training_set_clean.head()

Label                                                SMS  chinky  remembr  \
0   ham                  [yep, by, the, pretty, sculpture]       0        0   
1   ham  [yes, princess, are, you, going, to, make, me,...       0        0   
2   ham                    [welp, apparently, he, retired]       0        0   
3   ham                                           [havent]       0        0   
4   ham  [i, forgot, 2, ask, ü, all, smth, there, s, a,...       0        0   

   goigng  humans  pc  erupt  laugh  history  ...  hont  somewheresomeone  \
0       0       0   0      0      0        0  ...     0                 0   
1       0       0   0      0      0        0  ...     0                 0   
2       0       0   0      0      0        0  ...     0                 0   
3       0       0   0      0      0        0  ...     0                 0   
4       0       0   0      0      0        0  ...     0                 0   

   220  nor  boytoy  09066612661  09061743386  paining  suffers  sleepin  
0    0    0       0            0            0        0        0        0  
1    0    0       0            0            0        0        0        0  
2    0    0       0            0            0        0        0        0  
3    0    0       0            0            0        0        0        0  
4    0    0       0            0            0        0        0        0  

[5 rows x 7785 columns]

## Calculating Constants First

Now that we're done with data cleaning and have a training set to work with, we can begin creating the spam filter. Some of the terms in the four equations above will have the same value for every new message.

As a start, let's first calculate:

* P(Spam) and P(Ham)
* NSpam, NHam, NVocabulary

Attention:
* NSpam is equal to the number of words in all the spam messages — it's not equal to the number of spam messages, and it's not equal to the total number of unique words in spam messages.
* NHam is equal to the number of words in all the non-spam messages — it's not equal to the number of non-spam messages, and it's not equal to the total number of unique words in non-spam messages.
* We'll also use Laplace smoothing and set α=1.

In [14]:
# calculate P(Spam) and P(Ham)

training_set_clean['Label'].value_counts(normalize=True)

ham     0.86541
spam    0.13459
Name: Label, dtype: float64

In [15]:
p_spam=0.13459
p_ham=0.86541


In [16]:
# calculate N_Spam and N_Ham
spam_messages=training_set_clean[training_set_clean['Label']=='spam']
ham_messages=training_set_clean[training_set_clean['Label']=='ham']

#N_spam
n_words_per_spam_message=spam_messages['SMS'].apply(len)
n_spam=n_words_per_spam_message.sum()

#N_ham
n_words_per_ham_message = ham_messages['SMS'].apply(len)
n_ham = n_words_per_ham_message.sum()

#N_Vocabulary
n_vocabulary=len(vocabulary)

# Laplace smoothing
alpha = 1

## Calculating Parameters

Now that we have the constant terms calculated above, we can move on with calculating the parameters $P(w_i|Spam)$ and $P(w_i|Ham)$. Each parameter will thus be a conditional probability value associated with each word in the vocabulary.

The parameters are calculated using the formulas:

<img src="equation2_1.png">


In [17]:
# Initiate parameters
parameters_spam={unique_word: 0 for unique_word in vocabulary}
parameters_ham={unique_word: 0 for unique_word in vocabulary}

# Calculate parameters
for word in vocabulary:
    n_word_given_spam=spam_messages[word].sum()
    p_word_given_spam=(n_word_given_spam + alpha) / (n_spam + alpha*n_vocabulary)
    parameters_spam[word] = p_word_given_spam
    
    n_word_given_ham = ham_messages[word].sum()   
    p_word_given_ham = (n_word_given_ham + alpha) / (n_ham + alpha*n_vocabulary)
    parameters_ham[word] = p_word_given_ham

## Classifying A New Message
Now that we have all our parameters calculated, we can start creating the spam filter. The spam filter can be understood as a function that:

* Takes in as input a new message (w1, w2, ..., wn).
* Calculates P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn).
* Compares the values of P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn), and:
 * If P(Ham|w1, w2, ..., wn) > P(Spam|w1, w2, ..., wn), then the message is classified as ham.
 * If P(Ham|w1, w2, ..., wn) < P(Spam|w1, w2, ..., wn), then the message is classified as spam.
 * If P(Ham|w1, w2, ..., wn) = P(Spam|w1, w2, ..., wn), then the algorithm may request human help.


In [18]:
# create spam filter
import re

def spam_filter(message):
    message=re.sub('\W',' ', message)
    message = message.lower().split()
    
    p_spam_given_message=p_spam
    p_ham_given_message=p_ham
    
    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]

    print('P(Spam|message):', p_spam_given_message)
    print('P(Ham|message):', p_ham_given_message)
    
    if p_ham_given_message > p_spam_given_message:
        print('Label: Ham')
    elif p_ham_given_message < p_spam_given_message:
        print('Label: Spam')
    else:
        print('Equal proabilities, have a human classify this!')

In [19]:
#test the filter
spam_filter('WINNER!! This is the secret code to unlock the money: C3421.')

P(Spam|message): 1.3481340092074626e-25
P(Ham|message): 1.9368037883678308e-27
Label: Spam


## Test set
On the previous screen, we managed to create a spam filter, and we classified two new messages. We'll now try to determine how well the spam filter does on our test set of 1,114 messages.

The algorithm will output a classification label for every message in our test set, which we'll be able to compare with the actual label (given by a human). Note that, in training, our algorithm didn't see these 1,114 messages, so every message in the test set is practically new from the perspective of the algorithm.

In [20]:
def classify_test_set(message):    
    '''
    message: a string
    '''
    
    message = re.sub('\W', ' ', message)
    message = message.lower().split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
    
    if p_ham_given_message > p_spam_given_message:
        return 'ham'
    elif p_spam_given_message > p_ham_given_message:
        return 'spam'
    else:
        return 'needs human classification'

In [21]:
test_set['predicted'] = test_set['SMS'].apply(classify_test_set)
test_set.head()

Label                                                SMS predicted
0   ham          Later i guess. I needa do mcat study too.       ham
1   ham             But i haf enuff space got like 4 mb...       ham
2  spam  Had your mobile 10 mths? Update to latest Oran...      spam
3   ham  All sounds good. Fingers . Makes it difficult ...       ham
4   ham  All done, all handed in. Don't know if mega sh...       ham

## Accuracy of filter
Now, we'll write a function to measure the accuracy of our spam filter to find out how well our spam filter does.

In [24]:
correct=0
total=len(test_set)

for i, row in test_set.iterrows():
    if row['Label'] == row['predicted']:
        correct+=1
        
print('Correct:', correct)
print('Incorrect:', total - correct)
print('Accuracy:', correct/total)
    

Correct: 1100
Incorrect: 14
Accuracy: 0.9874326750448833


The accuracy is close to 98.74%, which is really good. Our spam filter looked at 1,114 messages that it hasn't seen in training, and classified 1,100 correctly.